core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java

```diff
@@ -14,40 +14,76 @@
 
 public class ExponentialGrowthPartitionMemoryEstimator
         implements PartitionMemoryEstimator
 {
+    private static final double ESTIMATION_QUANTILE = 0.9;
+
+    private final TDigest memoryUsageDistribution = new TDigest();
+
     @Override
     public MemoryRequirements getInitialMemoryRequirements(Session session, DataSize defaultMemoryLimit)
     {
         return new MemoryRequirements(
-                defaultMemoryLimit,
+                Ordering.natural().max(defaultMemoryLimit, getEstimatedMemoryUsage()),
                 false);
     }
 
     @Override
     public MemoryRequirements getNextRetryMemoryRequirements(Session session, MemoryRequirements previousMemoryRequirements, DataSize peakMemoryUsage, ErrorCode errorCode)
     {
         DataSize previousMemory = previousMemoryRequirements.getRequiredMemory();
-        DataSize baseMemory = Ordering.natural().max(peakMemoryUsage, previousMemory);
-        if (shouldIncreaseMemory(errorCode)) {
-            double growthFactor = SystemSessionProperties.getFaultTolerantExecutionTaskMemoryGrowthFactor(session);
-            return new MemoryRequirements(DataSize.of((long) (baseMemory.toBytes() * growthFactor), DataSize.Unit.BYTE), false);
+
+        // start with the maximum of previously used memory and actual usage
+        DataSize newMemory = Ordering.natural().max(peakMemoryUsage, previousMemory);
+        if (isOutOfMemoryError(errorCode)) {
+            // multiply if we hit an oom error
+            double growthFactor = getFaultTolerantExecutionTaskMemoryGrowthFactor(session);
+            newMemory = DataSize.of((long) (newMemory.toBytes() * growthFactor), DataSize.Unit.BYTE);
         }
-        return new MemoryRequirements(baseMemory, false);
+
+        // if we are still below current estimate for new partition let's bump further
+        newMemory = Ordering.natural().max(newMemory, getEstimatedMemoryUsage());
+
+        return new MemoryRequirements(newMemory, false);
     }
 
-    private boolean shouldIncreaseMemory(ErrorCode errorCode)
+    private boolean isOutOfMemoryError(ErrorCode errorCode)
     {
         return EXCEEDED_LOCAL_MEMORY_LIMIT.toErrorCode().equals(errorCode) // too many tasks from single query on a node
                 || CLUSTER_OUT_OF_MEMORY.toErrorCode().equals(errorCode); // too many tasks in general on a node
     }
+
+    @Override
+    public synchronized void registerPartitionFinished(Session session, DataSize peakMemoryUsage, boolean success, Optional<ErrorCode> errorCode)
+    {
+        if (success) {
+            memoryUsageDistribution.add(peakMemoryUsage.toBytes());
+        }
+        if (!success && errorCode.isPresent() && isOutOfMemoryError(errorCode.get())) {
+            double growthFactor = SystemSessionProperties.getFaultTolerantExecutionTaskMemoryGrowthFactor(session);
+            memoryUsageDistribution.add(peakMemoryUsage.toBytes() * growthFactor);
```

Comment:
I wonder if this should be something like `max(peakMemoryUsage.toBytes(), previousMemoryRequirements)`?

Reply:
Yeah - good point. We could have been killed because some other task gone crazy and filled in the pool.
I think there is no great answer here. As with that we may overshoot - but it is safe side so probably a way to go. And that is what we do in `getNextRetryMemoryRequirements` and this one should be in line with that. Will change.